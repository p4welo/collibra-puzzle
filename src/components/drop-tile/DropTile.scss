@import "../../styles/colors";
@import "../../styles/variables";

.DropTile {
  background: white;
  height: $tileSize - 2 * $borderWidth;
  width: $tileSize - 2 * $borderWidth;
  border: $borderWidth solid $blue;

  &--done {
    background: transparent;
  }

  &--hovered {
    background: yellow;
  }
}
