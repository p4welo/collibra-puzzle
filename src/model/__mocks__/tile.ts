import { Tile } from '../tile.model';

export const tileMock: Tile = {
  id: '0',
  done: false
};

export const doneTileMock: Tile = {
  id: '1',
  done: true
};

export const freshTilesMock: Tile[] = [
  { id: '0', done: false },
  { id: '1', done: false },
  { id: '2', done: false },
  { id: '3', done: false },
  { id: '4', done: false },
  { id: '5', done: false },
  { id: '6', done: false },
  { id: '7', done: false },
  { id: '8', done: false }
];

export const doneTilesMock: Tile[] = [
  { id: '0', done: true },
  { id: '1', done: true },
  { id: '2', done: true },
  { id: '3', done: true },
  { id: '4', done: true },
  { id: '5', done: true },
  { id: '6', done: true },
  { id: '7', done: true },
  { id: '8', done: true }
];

export const tileOrderMock: number[] = [
    0,1,2,3,4,5,6,7,8
];
