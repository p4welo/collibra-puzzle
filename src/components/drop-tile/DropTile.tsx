import React, { DragEvent, Component } from 'react';
import { Tile } from 'model/tile.model';
import './DropTile.scss';

interface DropTileProps {
  tile: Tile;
  finished: boolean;
  onDrop: (tileId: string) => void;
}

interface DropTileState {
  hovered: boolean
}

export class DropTile extends Component<DropTileProps, DropTileState> {

  state: DropTileState = {
    hovered: false
  };

  shouldComponentUpdate(nextProps: Readonly<DropTileProps>, nextState: Readonly<DropTileState>): boolean {
    return this.state.hovered !== nextState.hovered ||
        this.props.finished !== nextProps.finished ||
        this.props.tile !== nextProps.tile;
  }

  handleDrop = (e: DragEvent) => {
    e.preventDefault();
    const tileId = e.dataTransfer.getData('tile');
    this.setState({
      hovered: false
    });
    this.props.onDrop(tileId);
  };

  handleDragOver = (e: DragEvent) => {
    e.preventDefault();
    this.setState({
      hovered: true
    });
  };

  handleDragLeave = (e: DragEvent) => {
    e.preventDefault();
    this.setState({
      hovered: false
    });
  };

  get tileClass(): string {
    const isDone = this.props.tile.done;
    const isFinished = this.props.finished;
    const isHovered = !isDone && this.state.hovered;

    return `
      DropTile
      ${isDone ? 'DropTile--done' : ''}
      ${isFinished ? 'DropTile--finished' : ''}
      ${isHovered ? 'DropTile--hovered' : ''}
    `;
  }

  render() {
    return (
        <div className={this.tileClass}
            onDrop={this.handleDrop}
            onDragOver={this.handleDragOver}
            onDragLeave={this.handleDragLeave}
            data-testid='drop-tile'
        ></div>
    );
  }
}
