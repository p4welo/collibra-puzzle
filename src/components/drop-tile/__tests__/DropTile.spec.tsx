import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { tileMock, doneTileMock } from 'model/__mocks__/tile';
import { DropTile } from '../DropTile';

describe('DropTile component', () => {
  it('should have .DropTile--done class when tile is placed correctly', () => {
    const { getByTestId } = render(<DropTile tile={{ ...doneTileMock }} finished={false}
        onDrop={jest.fn()}/>);
    expect(getByTestId('drop-tile')).toHaveClass('DropTile--done');
  });

  it('should not have .DropTile--done class when tile is not placed correctly', () => {
    const { getByTestId } = render(<DropTile tile={{ ...tileMock }} finished={false}
        onDrop={jest.fn()}/>);
    expect(getByTestId('drop-tile')).not.toHaveClass('DropTile--done');
  });

  it('should have .DropTile--finished class when the game is finished', () => {
    const { getByTestId } = render(<DropTile tile={{ ...doneTileMock }} finished={true}
        onDrop={jest.fn()}/>);
    expect(getByTestId('drop-tile')).toHaveClass('DropTile--finished');
  });

  it('should have .DropTile--hovered class when is hovered by some tile and no done yet', () => {
    const { getByTestId } = render(<DropTile tile={{ ...tileMock }} finished={false}
        onDrop={jest.fn()}/>);
    fireEvent.dragOver(getByTestId('drop-tile'));
    expect(getByTestId('drop-tile')).toHaveClass('DropTile--hovered');

    fireEvent.dragLeave(getByTestId('drop-tile'));
    expect(getByTestId('drop-tile')).not.toHaveClass('DropTile--hovered');
  });

  it('should match snapshot when the game is finished', () => {
    const { container } = render(<DropTile tile={{ ...doneTileMock }} finished={true}
        onDrop={jest.fn()}/>);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should match snapshot when the game is not finished and tile is placed correctly', () => {
    const { container } = render(<DropTile tile={{ ...doneTileMock }} finished={false}
        onDrop={jest.fn()}/>);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should match snapshot when the game is not finished and tile is not placed correctly', () => {
    const { container } = render(<DropTile tile={{ ...tileMock }} finished={false}
        onDrop={jest.fn()}/>);
    expect(container.firstChild).toMatchSnapshot();
  });
});
