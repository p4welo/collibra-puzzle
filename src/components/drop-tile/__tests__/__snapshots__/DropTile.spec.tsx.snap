// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DropTile component should match snapshot when the game is finished 1`] = `
<div
  class="
      DropTile
      DropTile--done
      DropTile--finished
      
    "
  data-testid="drop-tile"
/>
`;

exports[`DropTile component should match snapshot when the game is not finished and tile is not placed correctly 1`] = `
<div
  class="
      DropTile
      
      
      
    "
  data-testid="drop-tile"
/>
`;

exports[`DropTile component should match snapshot when the game is not finished and tile is placed correctly 1`] = `
<div
  class="
      DropTile
      DropTile--done
      
      
    "
  data-testid="drop-tile"
/>
`;
